XmlNode.cpp

```diff
@@ -991,14 +991,15 @@ bool XmlNode::hasAttribute(const string &name)
 
 
 // ****************************************************************************
-/// Returns the integer value of the given attribute, or 0, if the attribute
-/// does not exist or is not convertable to integer.
+/// Returns the integer value of the given attribute, or an exception, if the attribute
```

Comment:
Exceptions are not returned but thrown. Please revise. Also state the type of exceptions explicitly (`std::invalid_argument`, `std::out_of_range`) so that caller can catch it appropriately. We also should differentiate between the `std::invalid_argument` exception thrown by `std::stoi` in case of conversion error and the `std::invalid_argument`. In fact, the optimal solution would be to create a custom exception like so (untested code):

```cpp
struct XmlException : public std::exception
{
    const char* what() const throw() override
    {
    	return "An XML exception occurred!";
    }
}

struct XmlAttributeNotFound : public XmlException
{
    const char* what() const throw() override
    {
    	return "Could not find attribute!";
    }
}

struct XmlBadAttribute : public XmlException
{
    const char* what() const throw() override
    {
    	return "Attribute could not be converted to expected type";
    }
}
```

That way we could throw `XmlAttributeNotFound` instead of `std::invalid_argument` and catch the `std::invalid_argument` from `std::stoi` and rethrow as `XmlBadAttribute`.

Reply:
std::stoi throw different messages than the other errors. You can already differentiate between them.

Reply:
Identifying an exception by its message is bad practice because it is implementation-dependent. Different compilers may throw different messages. The only safe (and also much more explicit) way is to use custom exceptions here.